esp-config/src/bin/esp-config/main.rs

```diff
@@ -0,0 +1,406 @@
+use std::{
+    collections::HashMap,
+    error::Error,
+    ops::Range,
+    path::{Path, PathBuf},
+};
+
+use clap::Parser;
+use env_logger::{Builder, Env};
+use esp_config::Value;
+use serde::Deserialize;
+use toml_edit::{DocumentMut, Formatted, Item, Table};
+use walkdir::WalkDir;
+
+mod tui;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Root of the project
+    #[arg(short = 'P', long)]
+    path: Option<PathBuf>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct CrateConfig {
+    name: String,
+    options: Vec<ConfigOption>,
+}
+
+// This partially mirrors [esp_config::ConfigOption]
+// but not exactly - `&'static str` is not great for deserializing
+// and this also serves as the data model for the TUI
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+pub struct ConfigOption {
+    name: String,
+    description: String,
+    default_value: Value,
+    actual_value: Value,
+    constraint: Option<Constraint>,
+    stability: Stability,
+    active: bool,
+}
+
+// This partially mirrors [esp_config::Stability]
+// but not exactly.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+pub enum Stability {
+    /// Unstable options need to be activated with the `unstable` feature
+    /// of the package that defines them.
+    Unstable,
+    /// Stable options contain the first version in which they were
+    /// stabilized.
+    Stable(String),
+}
+
+// This partially mirrors [esp_config::Validator]
+// but not exactly.
+// We especially can't and don't want to handle the `Custom` validator.
```

Comment:
Why can't we? What am I missing here :D?

Reply:
`Box<dyn Fn(&Value) -> Result<(), Error>>` isn't good for serialization/deserialization - and for having it clone,copy,eq etc.

(Fun fact: we don't use that validator at all currently IIRC - and while it's a wildcard for "everything" it is not nice to use it. 
I'd personally like to see it go away and instead add another validator whenever we see a need for it)

Reply:
Alternatively we could probably have a `Custom` validator using either a Rhai expression as we use it in other places or use the `evalexpr` crate which is a bit more lightweight

Reply:
If it's not used, let's not solve the problem we don't have